File: 2593-find-score-of-an-array-after-marking-all-elements/2593-find-score-of-an-array-after-marking-all-elements.cpp
class Solution {
public:
    #define P pair<int,int>
    long long findScore(vector<int>& nums) {
        int n = nums.size();
        priority_queue<P,vector<P>, greater<P>>minheap;
        vector<bool>visited(n,false);
        for(int i =0; i<n; i++){
            minheap.push({nums[i],i});
        }
        long long score =0;
        while(!minheap.empty()){
            P temp = minheap.top();
            int elem = temp.first, idx = temp.second;
            minheap.pop();

            if(!visited[idx]){
                visited[idx]=true;
                score+=elem;
                if(idx-1>=0 and !visited[idx-1]){
                    visited[idx-1]=true;
                }
                if(idx+1<n and !visited[idx+1]){
                    visited[idx+1]=true;
                }
            }
        }
        return score;
    }
};

File: 2593-find-score-of-an-array-after-marking-all-elements/2593-find-score-of-an-array-after-marking-all-elements.java
import java.util.PriorityQueue;
import java.util.Comparator;

class Solution {
    public long findScore(int[] nums) {
        int n = nums.length;
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            minHeap.offer(new int[]{nums[i], i});
        }

        long score = 0;

        while (!minHeap.isEmpty()) {
            int[] temp = minHeap.poll();
            int elem = temp[0], idx = temp[1];

            if (!visited[idx]) {
                visited[idx] = true;
                score += elem;

                if (idx - 1 >= 0 && !visited[idx - 1]) {
                    visited[idx - 1] = true;
                }
                if (idx + 1 < n && !visited[idx + 1]) {
                    visited[idx + 1] = true;
                }
            }
        }

        return score;
    }
}
